maple/__main__.py

```diff
@@ -0,0 +1,168 @@
+import argparse
+import code
+import csv
+from collections import defaultdict
+from datetime import datetime
+from pathlib import Path
+
+import pandas as pd
+from tqdm import tqdm
+
+from maple.classification import regex_classification
+from maple.db import connect
+from maple.types import Action, ActionType, Bill, Branch
+
+
+def parse_bills(bills_file: Path) -> list[Bill]:
+    with open(bills_file, "r") as f:
+        bill_actions = defaultdict(list)
+
+        for row in csv.DictReader(f, delimiter=","):
+            action = Action(
+                action=row["action"],
+                branch=Branch[row["branch"].lower()],
+                when=parse_datetime(row["date"]),
+            )
+            bill_actions[row["id"]].append(action)
+
+        return [Bill(id=k, history=v) for k, v in bill_actions.items()]
+
+
+def parse_datetime(dt: str) -> datetime:
+    return datetime.strptime(dt, "%Y-%m-%dT%H:%M:%S.%fZ")
+
+
+def load_command(args: argparse.Namespace) -> None:
+
+    with connect(args.db_path) as db:
+        for bill in tqdm(parse_bills(args.bills_file), unit="bills"):
+            db.add_bill(bill)
+
+
+def regex_command(args: argparse.Namespace) -> None:
+    with connect(args.db_path) as db:
+
+        ids = []
+        actions = []
+        predictions = []
+        labels = []
+
+        for id, action, label in db.actions_and_labels():
+            predicted = regex_classification(action)
+            ids.append(id)
+            actions.append(action.action)
+            predictions.append(predicted.value)
+            labels.append(label.value if label is not None else None)
+
+        df = pd.DataFrame(
+            data={
+                "action_id": ids,
+                "action": actions,
+                "prediction": predictions,
+                "label": labels,
+            }
+        )
+
+        if args.predictions_file is not None:
+            df.to_csv(args.predictions_file, index=False)
+
+        correct = len(df[df.prediction == df.label])
+        incorrect = len(df[(df.prediction != df.label) & ~pd.isna(df.label)])
+
+        print(f"{correct} correct predictions, {incorrect} incorrect predictions")
+
+        by_prediction = df.groupby("prediction").size().sort_values(ascending=False)
+        print()
+        print("Prediction counts:")
+        print(by_prediction)
+
+        types = []
+        labeleds = []
+        corrects = []
+        incorrects = []
+        precisions = []
+        recalls = []
+        f1s = []
+
+        for type in (x.value for x in ActionType):
```

Comment:
Not necessary, but -- you could do this with a `groupby` operation too, right?

Reply:
Almost certainly.